examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,605 @@
+):
+    # Load the pre-trained Xception model to be used as the base encoder.
+    xception = keras.applications.Xception(
+        include_top=False, weights="imagenet", pooling="avg"
+    )
+    # Set the trainability of the base encoder.
+    for layer in xception.layers:
+        layer.trainable = trainable
+    # Receive the images as inputs.
+    inputs = layers.Input(shape=(299, 299, 3), name="image_input")
+    # Preprocess the input image.
+    xception_input = tf.keras.applications.xception.preprocess_input(inputs)
+    # Generate the embeddings for the images using the xception model.
+    embeddings = xception(xception_input)
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the vision encoder model.
+    encoder = keras.Model(inputs, outputs, name="vision_encoder")
+    return encoder
+
+
+"""
+## Implement the text encoder
+
+In this example, we use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
+from [TensorFlow Hub](https://tfhub.dev) as the text encoder
+"""
+
+bert_module_url = (
+    "https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-512_A-8/1"
+)
+preprocess_model_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/2"
+
+
+def create_text_encoder(
+    num_projection_layers, projection_dims, dropout_rate, trainable=False
+):
+    # Load the BERT preprocessing module.
+    preprocess = hub.KerasLayer(preprocess_model_url, name="text_preprocessing")
+    # Load the pre-trained BERT model to be used as the base encoder.
+    bert = hub.KerasLayer(bert_module_url, "bert")
+    # Set the trainability of the base encoder.
+    bert.trainable = trainable
+    # Receive the text as inputs.
+    inputs = layers.Input(shape=(), dtype=tf.string, name="text_input")
+    # Preprocess the text.
+    bert_inputs = preprocess(inputs)
+    # Generate embeddings for the preprocessed text using the BERT model.
+    embeddings = bert(bert_inputs)["pooled_output"]
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the text encoder model.
+    encoder = keras.Model(inputs, outputs, name="text_encoder")
+    return encoder
+
+
+"""
+## Implement contrastive similarity loss
+
+To calculate the loss, we compute the pairwise dot-product similarity between
+each `caption_i` and `images_j` in the batch as the predictions.
+The target similarity between `caption_i`  and `image_j` is computed as
+the average of the (dot-product similarity between `caption_i` and `caption_j`)
+and (the dot-product similarity between `image_i` and `image_j`).
+Then we use crossentropy to compute the loss between the targets and the predictions.
+"""
+
+
+class ContrastiveSimilarityLoss(keras.losses.Loss):
```

Comment:
Given the fact that `y_true` is not used, I'm not sure a `Loss` subclass is the best option. A more natural factoring would be a custom `train_step` for the `dual_encoder` model. That said, the current factoring is still pretty simple, so I'll leave the decision up to you.